### PR TITLE
Gunpowder dataloading

### PR DESCRIPTION
**Pull request by Christianfoley**
Creating PR for preprocessing and training module rewrites using gunpowder.

What is left to be done:
- Alter inference so that it can be run without creating metadata from traditional preprocessing
- Clean and reorganize code as in issue #184

### CODE REVIEW

**Comment by Christianfoley** [comment]
Reopening and ready for review as of commit [`29f6f3d`](https://github.com/mehta-lab/microDL/pull/188/commits/29f6f3df9580ce9d55ad33befbf74fe01678731a) as consensus is to wait on merging [`inference_add_evaluation_metrics`](https://github.com/mehta-lab/microDL/tree/inference_add_evaluation_metrics) changes until after merge into `pytorch_implementation`.

Functionality is working for multi-position inference with features through #190, while features and changes beyond that (#197 through #203 ) will be added as PRs to the `pytorch_implementation`  or `main` branches.

**Comment by Soorya19Pradeep** [comment]
Hello @Christianfoley . I have been trying out the inference script and I ran into an error.

> Using GPU 0 with memory fraction 0.9841538595120257.
> PyTorch model load status: <All keys matched successfully>
> Traceback (most recent call last):
>   File "/home/soorya.pradeep/microDL/micro_dl/cli/torch_inference_script.py", line 103, in <module>
>     main(args.config, args.gpu, args.gpu_mem_frac)
>   File "/home/soorya.pradeep/microDL/micro_dl/cli/torch_inference_script.py", line 96, in main
>     torch_predictor.generate_dataloaders()
>   File "/home/soorya.pradeep/microDL/micro_dl/torch_unet/utils/inference.py", line 115, in generate_dataloaders
>     torch_data_container = ds.InferenceDatasetContainer(
>   File "/home/soorya.pradeep/microDL/micro_dl/torch_unet/utils/dataset.py", line 294, in __init__
>     self.train_dataset_list = self.generate_dataset_list(
>   File "/home/soorya.pradeep/microDL/micro_dl/torch_unet/utils/dataset.py", line 335, in generate_dataset_list
>     source_copy = source[i]
> TypeError: 'ZarrSource' object is not subscriptable

Am I missing something here from the earlier training step or something on the config file?

**Comment by Christianfoley** [comment]
Example config with new parameters for new inference module:
`/hpc/projects/comp_micro/projects/virtualstaining/torch_microDL/config_files/2022_11_01_VeroMemNuclStain/gunpowder_testing_12_13/`

**Comment by Soorya19Pradeep** [comment]
Hello @Christianfoley . Thank you for the reply. Here is the path to the config file which I have been using to run the inference:
`/hpc/projects/comp_micro/projects/infected_cell_imaging/VirtualStaining/config_files/torch_gpLoading_2023_02/torch_config_nucl25D.yml`

**Comment by Christianfoley** [comment]
> `/hpc/projects/comp_micro/projects/infected_cell_imaging/VirtualStaining/config_files/torch_gpLoading_2023_02/torch_config_nucl25D.yml`

This config file's zarr directory has not been preprocessed. There's another zarr dir called `Inputs_nucl` instead of `Inputs_Nucl` that HAS been preprocessed with unimodal masking. When I run inference with r2, corr, MAE, with `'unimodal'` as `mask_type` in the dataset it works without error. 

I sent the results to this folder since I don't have write permission in your model directory: `/hpc/projects/CompMicro/projects/virtual_staining/torch_microDL/models/2022_02_23_gunpowder_test/`
For completeness, here's the config I modified to use your data and model: `/hpc/projects/CompMicro/projects/virtualstaining/torch_microDL/config_files/2022_11_01_VeroMemNuclStain/gunpowder_testing_12_13/torch_config_Soorya_VeroA549_25D_nucl.yml`

File: micro_dl/inference/readme.md [hunk 1]
```diff
@@ -14,66 +14,38 @@ If there's not enough memory available on the GPU, and AssertionError will be ra
...
-    * num_slices (int): in case of 3D, the full volume will not fit in GPU
-    memory, specify the number of slices to use and this will depend on
-    the network depth, for ex 8 for a network of depth 4.
-    * inf_shape (list): Inference on a center sub volume.
-    * tile_shape (list): Shape of tile for tiling along xyz.
-    * num_overlap (int/list): int for tile_z, list for tile_xyz
-    * overlap_operation (str): e.g. 'mean'
-    
-## Metrics
+> **zarr_dir**: <span style="color:cyan"> absolute path </span> (absolute path to HCS-compatible zarr store containing data)
```
Comment (ziw-liu) [inline micro_dl/inference/readme.md:19]: @Christianfoley Are these html styles intended to be rendered on GitHub?

Comment (Christianfoley) [reply micro_dl/inference/readme.md:19]: Github does not render them, however they do render in VSCode. I'm not attached to them, but they don't affect the github rendering either way.

Comment (ziw-liu) [reply micro_dl/inference/readme.md:19]: I would discourage frequent html tags in markdown. In this case other native markup styles should be sufficient to distinguish these segments

**Pull request merged**